Sources/PenguinParallel/ThreadPool.swift

```diff
@@ -202,14 +164,32 @@ public struct InlineComputeThreadPool: TypedComputeThreadPool {
   /// Dispatch `fn` to be run at some point in the future (immediately).
   ///
   /// Note: this implementation just executes `fn` immediately.
-  public func dispatch(_ fn: (Self) -> Void) {
-    fn(self)
+  public func dispatch(_ fn: () -> Void) {
+    fn()
+  }
+
+  /// Executes `a` and `b` optionally in parallel, and returns when both are complete.
+  ///
+  /// Note: this implementation simply executes them serially.
+  public func join(_ a: () -> Void, _ b: () -> Void) {
```

Comment:
I think of `join` as an operation that waits on the completion of one or more already concurrently-executing things.

```suggestion
  public func concurrently(_ a: () -> Void, _ b: () -> Void) {
```

Just an idea.

Reply:
For context: I picked `join` as the [typical](https://en.wikipedia.org/wiki/Fork%E2%80%93join_model) [term-of-art](https://docs.rs/rayon/1.3.0/rayon/fn.join.html) in this space. I'm not fully sold on `concurrently` yet, because `join` represents _optional_ concurrency, which is important for performance at scale.

I think that it would be good to go over this API and think hard about naming & how the abstractions compose, but only once we understand the performance limitations & constraints. (Concretely, some of the (internal) abstractions are being re-written due to performance limitations in the current structure of things.)

Reply:
It *is* an established term-of-art, which means we should use it in the way that has been established.  You join stuff after you've forked it, in my experience, Rayon notwithstanding.  And IME when you fork, you're running the same closure in both threads of execution, with a parameter passed to indicate which invocation of the closure you got, similar to what you did with `parallelFor`, which I think might be better called `forkJoin` ersump'n.

I'm not sure the *concurrency* is optional from the programming model P.O.V., which is what matters here.  There may or may not be any actual *parallelism* between `a` and `b`'s execution, but that would be true even if you unconditionally launched separate threads for `a` and `b`, so surfacing that distinction as though it's significant seems like a mistake to me.

Also the word “optional” tends to imply it's up to the user, but it's not; this is up to the library.

As for putting off talking about naming and abstractions of “this API” until we know it's performance, I think if we don't do both at once, we don't know what “this API” *is*.  You don't want to design yourself into performance constraints based on assumptions about the programming model that don't actually apply.